File: quick_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm> // Required for std::swap

// Function to partition the array on the basis of pivot
int partition(std::vector<int>& arr, int low, int high) {
    int pivot = arr[high]; // Choosing the last element as pivot
    int i = (low - 1);     // Index of smaller element

    for (int j = low; j <= high - 1; j++) {
        // If current element is smaller than or equal to pivot
        if (arr[j] <= pivot) {
            i++; // Increment index of smaller element
            std::swap(arr[i], arr[j]);
        }
    }
    std::swap(arr[i + 1], arr[high]);
    return (i + 1);
}

// Function to implement Quick Sort
void quickSort(std::vector<int>& arr, int low, int high) {
    if (low < high) {
        // pi is partitioning index, arr[p] is now at right place
        int pi = partition(arr, low, high);

        // Separately sort elements before partition and after partition
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

// Utility function to print the array
void printArray(const std::vector<int>& arr) {
    for (int x : arr) {
        std::cout << x << " ";
    }
    std::cout << std::endl;
}

/*
// Example Usage:
int main() {
    std::vector<int> arr = {10, 7, 8, 9, 1, 5};
    int n = arr.size();
    std::cout << "Original array: ";
    printArray(arr);

    quickSort(arr, 0, n - 1);

    std::cout << "Sorted array: ";
    printArray(arr);

    std::vector<int> arr2 = {4, 2, 1, 3};
    int n2 = arr2.size();
    std::cout << "Original array 2: ";
    printArray(arr2);

    quickSort(arr2, 0, n2 - 1);

    std::cout << "Sorted array 2: ";
    printArray(arr2);

    return 0;
}
*/

File: quick_sort.py

from typing import List

def quick_sort(arr: List[int]) -> List[int]:
    """
    Sorts a list of integers in ascending order using the Quick Sort algorithm.

    Quick Sort is a highly efficient, comparison-based sorting algorithm.
    It works by selecting a 'pivot' element from the array and partitioning
    the other elements into two sub-arrays, according to whether they are
    less than or greater than the pivot. The sub-arrays are then sorted recursively.

    Time Complexity:
        - Best Case: O(n log n)
        - Average Case: O(n log n)
        - Worst Case: O(n^2) (occurs when the pivot selection consistently
          results in highly unbalanced partitions, e.g., already sorted array)

    Space Complexity:
        - O(log n) on average due to recursion stack.
        - O(n) in the worst case.

    Args:
        arr: A list of integers to be sorted.

    Returns:
        A new list containing the sorted integers.
    """
    if len(arr) <= 1:
        return arr
    else:
        pivot = arr[len(arr) // 2]
        left = [x for x in arr if x < pivot]
        middle = [x for x in arr if x == pivot]
        right = [x for x in arr if x > pivot]
        return quick_sort(left) + middle + quick_sort(right)

if __name__ == "__main__":
    # Test cases
    test_arrays = [
        [],
        [1],
        [5, 2, 9, 1, 7],
        [3, 3, 3, 3, 3],
        [10, 9, 8, 7, 6, 5, 4, 3, 2, 1],
        [-5, 0, 5, -10, 10],
        [1, 2, 3, 4, 5]
    ]

    for i, arr in enumerate(test_arrays):
        sorted_arr = quick_sort(arr)
        print(f"Original array {i+1}: {arr}")
        print(f"Sorted array {i+1}: {sorted_arr}")
        print("-" * 30)
